src/rust/lib/generics/src/tuple.rs

```diff
@@ -0,0 +1,563 @@
+        (t0,)
+    }
+}
+
+impl<T0,T1> Into<(T0,T1)>
+for hlist::ty![T0,T1] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1) {
+        let hlist::pat![t0,t1] = self;
+        (t0,t1)
+    }
+}
+
+impl<T0,T1,T2> Into<(T0,T1,T2)>
+for hlist::ty![T0,T1,T2] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2) {
+        let hlist::pat![t0,t1,t2] = self;
+        (t0,t1,t2)
+    }
+}
+
+impl<T0,T1,T2,T3> Into<(T0,T1,T2,T3)>
+for hlist::ty![T0,T1,T2,T3] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3) {
+        let hlist::pat![t0,t1,t2,t3] = self;
+        (t0,t1,t2,t3)
+    }
+}
+
+impl<T0,T1,T2,T3,T4> Into<(T0,T1,T2,T3,T4)>
+for hlist::ty![T0,T1,T2,T3,T4] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3,T4) {
+        let hlist::pat![t0,t1,t2,t3,t4] = self;
+        (t0,t1,t2,t3,t4)
+    }
+}
+
+impl<T0,T1,T2,T3,T4,T5> Into<(T0,T1,T2,T3,T4,T5)>
+for hlist::ty![T0,T1,T2,T3,T4,T5] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3,T4,T5) {
+        let hlist::pat![t0,t1,t2,t3,t4,t5] = self;
+        (t0,t1,t2,t3,t4,t5)
+    }
+}
+
+impl<T0,T1,T2,T3,T4,T5,T6> Into<(T0,T1,T2,T3,T4,T5,T6)>
+for hlist::ty![T0,T1,T2,T3,T4,T5,T6] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3,T4,T5,T6) {
+        let hlist::pat![t0,t1,t2,t3,t4,t5,t6] = self;
+        (t0,t1,t2,t3,t4,t5,t6)
+    }
+}
+
+impl<T0,T1,T2,T3,T4,T5,T6,T7> Into<(T0,T1,T2,T3,T4,T5,T6,T7)>
+for hlist::ty![T0,T1,T2,T3,T4,T5,T6,T7] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3,T4,T5,T6,T7) {
+        let hlist::pat![t0,t1,t2,t3,t4,t5,t6,t7] = self;
+        (t0,t1,t2,t3,t4,t5,t6,t7)
+    }
+}
+
+impl<T0,T1,T2,T3,T4,T5,T6,T7,T8> Into<(T0,T1,T2,T3,T4,T5,T6,T7,T8)>
+for hlist::ty![T0,T1,T2,T3,T4,T5,T6,T7,T8] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3,T4,T5,T6,T7,T8) {
+        let hlist::pat![t0,t1,t2,t3,t4,t5,t6,t7,t8] = self;
+        (t0,t1,t2,t3,t4,t5,t6,t7,t8)
+    }
+}
+
+impl<T0,T1,T2,T3,T4,T5,T6,T7,T8,T9> Into<(T0,T1,T2,T3,T4,T5,T6,T7,T8,T9)>
+for hlist::ty![T0,T1,T2,T3,T4,T5,T6,T7,T8,T9] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3,T4,T5,T6,T7,T8,T9) {
+        let hlist::pat![t0,t1,t2,t3,t4,t5,t6,t7,t8,t9] = self;
+        (t0,t1,t2,t3,t4,t5,t6,t7,t8,t9)
+    }
+}
+
+impl<T0,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10> Into<(T0,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10)>
+for hlist::ty![T0,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10) {
+        let hlist::pat![t0,t1,t2,t3,t4,t5,t6,t7,t8,t9,t10] = self;
+        (t0,t1,t2,t3,t4,t5,t6,t7,t8,t9,t10)
+    }
+}
+
+impl<T0,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11> Into<(T0,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11)>
+for hlist::ty![T0,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11] {
+    #[inline(always)]
+    fn into(self) -> (T0,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11) {
+        let hlist::pat![t0,t1,t2,t3,t4,t5,t6,t7,t8,t9,t10,t11] = self;
+        (t0,t1,t2,t3,t4,t5,t6,t7,t8,t9,t10,t11)
+    }
+}
```

Comment:
Cannot we just automatically implement Push and Pop traits for everything we can convert to HList and from Pop/Front::Output?

All operations could be done by moving, so they shan't do any performance drop (compiler should handle it quite smoothly.

Reply:
I dont remember why but lets not fix it now, its complex. Im adding a todo comment there